identity-iota/src/did/doc/iota_document.rs

```diff
@@ -67,38 +67,30 @@ impl TryMethod for IotaDocument {
 }
 
 impl IotaDocument {
-  /// Creates a new DID Document from the given KeyPair.
-  ///
-  /// The DID Document will be pre-populated with a single authentication
-  /// method based on the provided [KeyPair].
-  ///
-  /// The authentication method will have the DID URL fragment `#authentication`
-  /// and can be easily retrieved with [Document::authentication].
-  pub fn from_keypair(keypair: &KeyPair) -> Result<Self> {
-    let method: IotaVerificationMethod = IotaVerificationMethod::from_keypair(keypair, "authentication")?;
-
-    // SAFETY: We don't create invalid Methods.  Method::from_keypair() uses the MethodBuilder
-    // internally which verifies correctness on construction.
-    Ok(unsafe { Self::from_authentication_unchecked(method) })
-  }
+  const DEFAULT_METHOD_FRAGMENT: &'static str = "authentication";
 
-  /// Creates a new DID Document from the given KeyPair and network.
+  /// Creates a new DID Document from the given [`KeyPair`] and network, defaulting to
+  /// [`Network::Mainnet`](crate::tangle::Network::Mainnet) if unspecified.
   ///
-  /// The DID Document will be pre-populated with a single authentication
-  /// method based on the provided [KeyPair].
+  /// The DID Document will be pre-populated with a single authentication verification method
+  /// derived from the provided [`KeyPair`]. This method will have the DID URL fragment
+  /// `#authentication` and can be easily retrieved with [`Document::authentication`].
   ///
-  /// The authentication method will have the DID URL fragment `#authentication`
-  /// and can be easily retrieved with [Document::authentication].
-  pub fn from_keypair_with_network(keypair: &KeyPair, network: &str) -> Result<Self> {
-    let method: IotaVerificationMethod =
-      IotaVerificationMethod::from_keypair_with_network(keypair, "authentication", network)?;
-
-    // SAFETY: We don't create invalid Methods.  Method::from_keypair() uses the MethodBuilder
-    // internally which verifies correctness on construction.
-    Ok(unsafe { Self::from_authentication_unchecked(method) })
+  /// NOTE: the generated document is unsigned, see [`Document::sign`].
+  pub fn new(keypair: &KeyPair, network: Option<&str>) -> Result<Self> {
```

Comment:
I'm not a huge fan of the optional argument. I think having two methods `new` and `new_with_network` would be preferable. Seems to be more in line with the rest of the library and more idiomatic for Rust. What do you think?

Reply:
I'm glad you brought attention to this, I originally chose a single `new` constructor for a few reasons:

1. to move away from the `from_x` naming scheme to `new`, which is more idiomatic for constructors,
2. for clarity by reducing the number of options a developer has when constructing a `Document`,
3. to make it clear that `network` is important to specify for a new `Document`.

Previously we had `from_keypair`, `from_keypair_with_network` and `from_authentication` (also `from_authentication_unchecked`, `try_from_core`, and `try_from_base` but those are more clearly not the intended constructors). It's far more clear with just `new` and `from_authentication` which the intended method is.

My arguments in favour of `new` and `new_with_network`:

- it is similarly more idiomatic than `from_x` 
- adding new constructor options would be a non-breaking change (e.g. `new_with_network_and_tag` would be a separate function)
- it fits the convention in the code base (or at least aligns with the current `from_x`, `from_x_with_y` scheme)
- creating a new `Document` on the `mainnet` is simpler (`new(key)` vs `new(key, None)`)

My arguments against using `new` and `new_with_network` are:

- many-constructors problem: with just two fields `key` and `network` it's fine, but if we want to add the default `tag`/`fragment` we could introduce `new_with_tag` and `new_with_network_and_tag` too.
- longer constructor names: `new_with_network_and_tag` is probably the upper-limit of what I would tolerate
- nit: brevity in the external API for the low-level API is not as important as the `Account`, particularly if we're intending on pushing towards the `Account` as the default.

For the sake of completeness, the alternatives would be either the `Builder` pattern, or a sort of `Config` pattern where one creates a single configuration struct as input to the `Document` constructor, and that can be extended at-will (but this just "moves" the construction problem to the `Config` struct). I consider the `Builder` pattern overkill for 2-3 fields and a `DocumentConfig` an anti-pattern in this case.

**tl;dr** Overall I am slightly in favour of `new` and `new_with_network`. However, I also think developers should be able to specify the fragment of the initial verification method somehow. So the only real question is how do we handle that?

**Option 1**:

- `new(keypair, network, fragment)`

**Option 2**:

- `new(keypair)`
- `new_with_network(keypair, network, fragment)`

**Option 3**:

- `new(keypair)`
- `new_with_network(keypair, network)`
- `new_with_network_and_fragment(keypair, network, fragment)`

**Option 4**:

Don't allow setting the default verification method fragment on creation. Developers will have to change the fragment for both the verification method and authentication reference, which is a chore but not impossible.

Reply:
Thanks a lot for the write-up! I like the overall change to move away from `from_*` to the more idiomatic `new*`. I think the fragment should be specifiable, since it's the low-level API which should give control over something like that. I also agree that builder and `DocumentConfig` would be overkill.

I feel like option 2 would be a good middle ground. I'm not sure what signature you had in mind for `new_with_network`, but we could make this the constructor with optional arguments. So keypair is required, but both `network` and `fragment` are `Option<&str>`. To reflect in its name that it offers two optional fields, we could go with `new_with_options`, or something like that - can't think of anything better right now. If one wants to create a document with the default network but a custom fragment, then `IotaDocument::new_with_options(keypair, None, Some("fragment"));` seems acceptable.

Reply:
I like `new_with_options`!

Reply:
`new_with_options` is a good suggestion and probably the convention I would use going forward: `new()` with required fields for an easy-to-use default constructor, `new_with_options()` to cover everything else except where we use the `Builder` pattern.